syntax = "proto3";

package spice_server;

import "spice.proto";

// Predefined Spice simulators, which must be installed and configured on the
// server. Not all versions might be available.
enum Flavour {
  UNSET = 0;

  XYCE = 1;
  NGSPICE = 2;
  SPECTRE = 3;
  HSPICE = 4;

  // RESERVED 4 - 100.

  XYCE_7_8 = 1001;
  XYCE_7_9 = 1002;
  XYCE_7_10 = 1003;
}

message FileInfo {
  string path = 1;
  bytes data = 2;
}

message SimulatorInfo {
  string name = 1;

  string version = 2;

  // The flavours which refer to this Simulator.
  repeated Flavour flavours = 3;

  string license = 4;

  // Maybe.
  string path = 5;
}

message ListSimulatorsRequest {
}

message ListSimulatorsResponse {
  repeated SimulatorInfo simulators = 1;
}

message VerbatimFileInput {
  repeated FileInfo files = 1;
}

// Request to run a SPICE simulation
message SimulationRequest {
  // Simulator to use (e.g., "ngspice", "ltspice", "xyce")
  Flavour simulator = 1;

  // The netlist content to simulate
  // 
  // Either the user specified everything in abstract VLSIR protobufs, or they
  // provide the input files to provide to the simulator verbatim.
  //
  // TODO(aryap): I think how this works is that the first file is provided to
  // the simulator and remaining files and put in their relative places in the
  // temporary directory.
  //
  // I imagine that accepting raw files is a security risk that may be
  // disallowed on some servers, so it will probably be easier to create a
  // separate RPC endpoint for that altogether.
  oneof inputs {
    vlsir.spice.SimInput vlsir_sim_input = 2;
    VerbatimFileInput verbatim_files = 3;
  }

  // Additional simulator arguments.
  repeated string additional_args = 10;
}

// Streaming response containing simulation output
message SimulationResponse {
  // Output line from the simulator
  string output = 1;

  // Whether this is stdout or stderr
  enum StreamType {
    STDOUT = 0;
    STDERR = 1;
  }
  StreamType stream_type = 2;

  // Exit code (only set in final message)
  optional int32 exit_code = 3;

  

  // Whether this is the final message
  bool done = 4;
}

// SpiceServer service definition
service SpiceSimulator {
  // Run a SPICE simulation and stream results back
  rpc RunSimulation(SimulationRequest) returns (stream SimulationResponse);
  rpc ListSimulators(ListSimulatorsRequest) returns (ListSimulatorsResponse);
}
